Create Zone Hold functionality

danquack opened the pull request:
## Description

Adds the ability to create a zone hold based on the hold API.

## Has your change been tested?

I called the apis manually. I am happy to write tests, but I'm not sure where to get started. I could stub the responses, but since this is basically just calling the raw api, im not sure what I would be testing (no additional logic).

## Screenshots (if appropriate):

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented in [cloudflare/api-schemas](https://github.com/cloudflare/api-schemas) 
      and relies on stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/
closes #1364

danquack committed:
Create Zone Hold functionality

danquack committed:
changelog

patryk commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+	"github.com/goccy/go-json"
+)
+
+// ZoneHold represents a Zone Hold rule. A rule only permits access to
```
The comment doesn't match the expected behaviour. You may use copy found on https://developers.cloudflare.com/api/operations/zones-0-hold-post.

danquack replied:
A lot of these were copy pasta from a different file. I updated not, can you check if that works for you.

patryk commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+// for creating a zone Hold.
+type ZoneHoldCreateParams struct {
+	ID                string `json:"id"`
+	Hold              bool   `json:"hold"`
```
These two fields are not defined for POST request.

patryk commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+// for updating a zone Hold.
+type ZoneHoldDeleteParams struct {
+	ID        string  `json:"id"`
+	Hold      bool    `json:"hold"`
```
Same.

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+	ResultInfo `json:"result_info"`
+}
+
+// ZoneHoldDeleteResponse represents a response from the List Zone Hold
```
Comment describes a List Zone Hold endpoint, which does not exist.

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+	Result ZoneHold `json:"result"`
+}
+
+// ZoneHoldCreateParams contains required and optional params
```
There is only 1 parameter (`include_subdomains`) for Create Zone Hold Request, and it is optional.

danquack committed:
Update comments

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+	IncludeSubdomains *bool  `json:"include_subdomains,omitempty"`
+}
+
+// ZoneHoldDeleteParams contains required and optional params
```
There is only 1 parameter (`hold_after`) for a Delete Zone Hold request and it is optional.

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+type ZoneHoldDeleteParams struct {
+	ID        string  `json:"id"`
+	Hold      bool    `json:"hold"`
+	HoldAfter *string `json:"hold_after,omitempty"`
```
This must be an RFC3339-formatted timestamp, should the type be changed to `time` and this conversion done as part of the Delete Operation to ensure operation performs as expected?

danquack replied:
converted.

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+func (api *API) CreateZoneHold(ctx context.Context, zoneID string, params ZoneHoldCreateParams) (ZoneHold, error) {
+	uri := fmt.Sprintf("/zones/%s/hold", zoneID)
+	if params.IncludeSubdomains != nil {
+		uri = fmt.Sprintf("%s?include_subdomains=%t", uri, *params.IncludeSubdomains)
```
Thoughts on using `url` package's `Values` type for this encoding?

https://stackoverflow.com/a/36000522/8156448

jacobbednarz replied:
we have [`buildURI`](https://github.com/cloudflare/cloudflare-go/blob/80b7949c7f0b8f051b7b5e79f6272f30c552c00c/utils.go#L13-L16) for this purpose. it allows adding query parameter tags to the struct without worrying about manually building path + query parameters.

example: https://github.com/cloudflare/cloudflare-go/blob/80b7949c7f0b8f051b7b5e79f6272f30c552c00c/tunnel_routes.go#L85

tpaulus commented on hold.go:
```diff
@@ -0,0 +1,112 @@
+func (api *API) DeleteZoneHold(ctx context.Context, zoneID string, params ZoneHoldDeleteParams) (ZoneHold, error) {
+	uri := fmt.Sprintf("/zones/%s/hold", zoneID)
+	if params.HoldAfter != nil {
+		uri = fmt.Sprintf("%s?hold_after=%s", uri, *params.HoldAfter)
```
Thoughts on using `url` package's `Values` type for this encoding?

https://stackoverflow.com/a/36000522/8156448

danquack replied:
Id be fine either way, my thought to use sprintf was just because it was a total of one param

danquack committed:
remove unused params

danquack committed:
string -> time.Time

jacobbednarz commented on hold.go:
```diff
@@ -0,0 +1,109 @@
+// Enforce a zone hold on the zone, blocking the creation and activation of zones with this zone's hostname.
+//
+// API reference: https://developers.cloudflare.com/api/operations/zones-0-hold-post
+func (api *API) CreateZoneHold(ctx context.Context, zoneID string, params ZoneHoldCreateParams) (ZoneHold, error) {
```
we need to update the method signature here to match the library conventions - https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods

danquack replied:
Questions on the ResourceContainer. I copied my functions from here? Are they outside convention?

https://github.com/cloudflare/cloudflare-go/blob/64904badcb5202c3ad02d4a1c90bbfa7432f9636/firewall.go#L117

if I switch to the container, where does zone ID come from? Does it come from the container or do I have to add it as a param?

jacobbednarz replied:
yes, that is an older method that has not yet been converted to the patterns outlined in the conventions and [experimental client doc](https://github.com/cloudflare/cloudflare-go/blob/master/docs/experimental.md). you're best off checking out a more modern file like https://github.com/cloudflare/cloudflare-go/blob/master/rulesets.go#L758 to see how it works.

> if I switch to the container, where does zone ID come from? Does it come from the container or do I have to add it as a param?

`*ResourceContainer` is the basic way of describing a resource and its identifier (https://github.com/cloudflare/cloudflare-go/blob/master/resource.go). the method accepts this as a parameter but end usage should use the helper method like [`ZoneIdentifier`](https://github.com/cloudflare/cloudflare-go/blob/a0f7a6b7de3d9023ef32fa9f518b94e4d68e3317/resource.go#L99-L105) which handles creating the struct and the associated relationships for you.

jacobbednarz commented on hold.go:
```diff
@@ -0,0 +1,109 @@
+// Retrieve whether the zone is subject to a zone hold, and metadata about the hold.
+//
+// API reference: https://developers.cloudflare.com/api/operations/zones-0-hold-get
+func (api *API) ZoneHold(ctx context.Context, zoneID string) (ZoneHold, error) {
```
updating this to follow the library  conventions for CRUD naming

```suggestion
func (api *API) GetZoneHold(ctx context.Context, zoneID string) (ZoneHold, error) {
```

we'll also need to update the signature to match https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods

jacobbednarz reviewed (changes requested):
thanks for the PR! 

i've left some feedback inline about the library conventions we need to update here but we'll also need to add test coverage. you can check out another file like https://github.com/cloudflare/cloudflare-go/blob/master/accounts_test.go to get an idea on the minimums we expect here.

danquack committed:
Apply suggestions from code review

Co-authored-by: Jacob Bednarz <jacob.bednarz@hey.com>

jacobbednarz committed:
rename `hold` to `zone_hold` to better define it's contents

jacobbednarz committed:
add zone holds test cases

Signed-off-by: Jacob Bednarz <jacob.bednarz@gmail.com>

jacobbednarz committed:
swap zone hold methods to using library conventions for method signatures

Signed-off-by: Jacob Bednarz <jacob.bednarz@gmail.com>

jacobbednarz committed:
remove dedicated `ZoneHoldDeleteResponse`

The response body is identical to other zone hold responses so let's reuse that instead.

Signed-off-by: Jacob Bednarz <jacob.bednarz@gmail.com>
